Support running GeoPartitioned TPCC

deeps1991 opened the pull request:
This patch provides support to optionally run TPCC in a geopartitioned
setup by partitioning all the TPCC tables by warehouse-id. This will
allow placement of data such that all the rows in all the tables
associated with a particular range of warehouses will always be
located in a particular zone.

Specifying the number of partitions, and the placement configuration
for each partition can be provided in XML format in geopartitioned_workload.xml

GeoPartitioned TPCC is different from the regular TPCC in the
following ways:
1) The tables needed for TPCC are all created in a partitioned
manner. We use YSQL Declarative Partitioning to split the total
number of warehouses into a set of ranges, and assign these ranges
to partitions. Each TPCC table is split by warehouse ID in the same
way, thus all the data pertaining to a warehouse can be found in
the same zone.

2) A Foreign key reference from tableX->tableY is implemented as
tableX_part1-->tableY_part1, tableX_part2-->tableY_part2...

3) SQL functions do not support partition pruning, unlike PL/Pgsql
Procedures. However SQL functions support batching of writes. Hence
to have the best of both worlds, the updatestock functions now
calculates the appropriate partition of the table to update and updates
the appropriate partition table (If we had used Pl/Pgsql procedures, we
could have had the updatestock function update the partitioned table
directly, and the planner would have determined the right partition to
update. However this would have removed the ability to batch updates).

4) 15% of Payment transactions and 1% of NewOrder transactions operate
on a "remote warehouse". Ordinarily this remote warehouse would be a
random number picked across all warehouses. However, if geopartitioning
is enabled, the "remote warehouse" is a random number picked within the
same partition as the "local" warehouse. This is a temporary restriction
until YSQL can support Foreign Keys on partitioned tables.

deeps1991 committed:
Support running GeoPartitioned TPCC

This patch provides support to optionally run TPCC in a geopartitioned
setup by partitioning all the TPCC tables by warehouse-id. This will
allow placement of data such that all the rows in all the tables
associated with a particular range of warehouses will always be
located in a particular zone.

Specifying the number of partitions, and the placement configuration
for each partition can be provided in XML format in
geopartitioned_workload.xml

GeoPartitioned TPCC is different from the regular TPCC in the
following ways:
1) The tables needed for TPCC are all created in a partitioned
manner. We use YSQL Declarative Partitioning to split the total
number of warehouses into a set of ranges, and assign these ranges
to partitions. Each TPCC table is split by warehouse ID in the same
way, thus all the data pertaining to a warehouse can be found in
the same zone.

2) A Foreign key reference from tableX->tableY is implemented as
tableX_part1-->tableY_part1, tableX_part2-->tableY_part2...

3) SQL functions do not support partition pruning, unlike PL/Pgsql
Procedures. However SQL functions support batching of writes. Hence
to have the best of both worlds, the updatestock functions now
calculates the appropriate partition of the table to update and updates
the appropriate partition table (If we had used Pl/Pgsql procedures, we
could have had the updatestock function update the partitioned table
directly, and the planner would have determined the right partition to
update. However this would have removed the ability to batch updates).

4) 15% of Payment transactions and 1% of NewOrder transactions operate
on a "remote warehouse". Ordinarily this remote warehouse would be a
random number picked across all warehouses. However, if geopartitioning
is enabled, the "remote warehouse" is a random number picked within the
same partition as the "local" warehouse. This is a temporary restriction
until YSQL can support Foreign Keys on partitioned tables.

robertsami commented on src/com/oltpbenchmark/WorkloadConfiguration.java:
```diff
@@ -358,6 +360,19 @@ public void setMaxLoaderRetries(int maxLoaderRetries) {
+  }
+
+  public boolean getGeoPartitioningEnabled() {
+      return geoPartitionPolicy.isPartitioningEnabled();
```
is geoPartitionPolicy possibly null?

deeps1991 replied:
No, it is always initialized with whether geopartitioning is enabled or not.

d-uspenskiy replied:
@deeps1991 

In this case it is better to remove setter function `setGeoPartitionPolicy` and configure `geoPartitionPolicy` field in constructor

deeps1991 replied:
Done

robertsami commented on src/com/oltpbenchmark/api/Procedure.java:
```diff
@@ -95,6 +97,17 @@ public final InstrumentedPreparedStatement getPreparedStatement(Connection conn,
+        int partitionIndex = (warehouseId - 1) / policy.numWareHousesPerSplit();
+        int startPartition = partitionIndex * policy.numWareHousesPerSplit() + 1;
+        int endPartition = (partitionIndex + 1) * policy.numWareHousesPerSplit();
+        return TPCCUtil.randomNumber(startPartition, endPartition, gen);
```
please comment on what we're doing here and the motivation

robertsami commented on src/com/oltpbenchmark/api/Procedure.java:
```diff
@@ -95,6 +97,17 @@ public final InstrumentedPreparedStatement getPreparedStatement(Connection conn,
         return (this.procName);
     }
+    
+    public int getRandomWarehouseId(Worker w, int warehouseId, int numWarehouses, Random gen) {
```
can this live in TPCCUtil as well rather than in Procedure? it's good to keep all the random number generation code together when possible as this code should conform to the same section in the tpcc spec

deeps1991 replied:
Done

robertsami commented on src/com/oltpbenchmark/api/Loader.java:
```diff
@@ -160,7 +160,7 @@ public void load(Connection conn) {
 
     private PreparedStatement getInsertStatement(Connection conn, String tableName) throws SQLException {
-      return conn.prepareStatement(SchemaManager.getInsertDml(tableName));
+      return conn.prepareStatement(DefaultSchemaManager.getInsertDml(tableName));
```
should this be always using DefaultSchemaManager? this is a bit confusing. if we want to use the same getInsertDml logic in both cases, we should at least park the static method in some class that is not semantically specific about the use case (i.e. DefaultSchemaManager semantically specifies it should be used in the non-geo-partitioned case, so the reader could think this is a mistake)

deeps1991 replied:
Nice catch. I moved it to TPCCTableSchema instead

robertsami commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,47 @@
+package com.oltpbenchmark.util;
+
+public class PlacementBlock {
```
please add a descriptive comment to this class. also any other classes you introduce in this PR. i just noticed and thought to comment on it here because its actually not clear what this does at a glance

deeps1991 replied:
Done

robertsami commented on src/com/oltpbenchmark/api/TPCCTableSchemas.java:
```diff
@@ -0,0 +1,147 @@
+import com.oltpbenchmark.benchmarks.tpcc.TPCCConstants;
+
+public class TPCCTableSchemas {
+    public static final Map<String, TableSchema> tables = Stream.of(
```
please fix indenting of these builders. would be easier to glance at if the `.foo()` calls below the initial line of each builder were indented

deeps1991 replied:
Done

robertsami commented on src/com/oltpbenchmark/api/BenchmarkModule.java:
```diff
@@ -393,36 +392,112 @@ public void enableForeignKeys() throws Exception {
-            st.execute(updateStmt);
-          }
+    // Stock table, and a procedure to get stock levels of items recently ordered.
+    public void createSqlProcedures() throws Exception {
```
maybe its time to expose a `createSqlProcedures` method on `SchemaManager` and forward these calls to a `BenchmarkModule` owned instance of a `SchemaManager` generated with the `workConf`? then each base class can implement the appropriate behavior, opaque to this class, and that logic lives with other schema specification code

deeps1991 replied:
Very good point! Done.

robertsami reviewed (changes requested):
high level things:
1. already spoke offline about seeing if it makes sense to separate out geo-partitioned config into new file, and parsing logic to new file as well
2. i wonder if we could add a directory `src/com/oltpbenchmark/schema/` and park all the interfaces, base classes, or generalize-able classes in there, and the deployment strategy specific classes in `src/com/oltpbenchmark/schema/default` and `src/com/oltpbenchmark/schema/geopartitioned` respectively?

deeps1991 commented:
Hey @robertsami Thanks for such a detailed review! I addressed all your comments in this patch, please let me know if you have more.

d-uspenskiy commented on src/com/oltpbenchmark/schema/Column.java:
```diff
@@ -0,0 +1,22 @@
+ * Helper class denoting a column in a table in the TPCC schema.
+ */
+public class Column {
+    final String name;
```
Why no `private`?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/GeoPartitionedConfigFileOptions.java:
```diff
@@ -0,0 +1,134 @@
+            final String name = getRequiredStringOpt(path + "name");
+
+            Optional<Boolean> isStorePartitionedTables = getBoolOpt(path + "storePartitionedTables");
+            if (isStorePartitionedTables.isPresent() && isStorePartitionedTables.get()) {
```
It seems like helper function `bool FlagIsTrue(Optional<Boolean> flag)` will help to make code more readable

deeps1991 replied:
Nice point, done

d-uspenskiy commented on src/com/oltpbenchmark/GeoPartitionedConfigFileOptions.java:
```diff
@@ -0,0 +1,134 @@
+            }
+
+            PlacementPolicy placementPolicy = new PlacementPolicy();
+            final int totalReplicas = getIntOpt(path + "replicationFactor").get().intValue();
```
Is the `intValue()` necessary here?

deeps1991 replied:
Turns out, not. Done.

d-uspenskiy commented on src/com/oltpbenchmark/GeoPartitionedConfigFileOptions.java:
```diff
@@ -0,0 +1,134 @@
+              LOG.error(String.format("Invalid Geoplacement policy: The sum of minReplicationFactor %d" +
+                    " is greater than total replication factor %d", totalMinReplicationFactor,
+                    totalReplicas));
+              System.exit(1);
```
Is it good idea to use `System.exit(1)` in multiple places? Is it possible to throw exception, catch them in one place (in main() function for example) and make required logging

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultTable.java:
```diff
@@ -0,0 +1,47 @@
+
+public class DefaultTable extends Table {
+    
+    private String tablespace = null;
```
Maybe `final`?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/GeoPartitionedSchemaManager.java:
```diff
@@ -0,0 +1,190 @@
+
+    private static final Logger LOG = Logger.getLogger(GeoPartitionedSchemaManager.class);
+
+    private GeoPartitionPolicy geoPartitionPolicy;
```
Maybe `final`?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/GeoPartitionedSchemaManager.java:
```diff
@@ -0,0 +1,190 @@
+
+    private GeoPartitionPolicy geoPartitionPolicy;
+
+    private Map<String, Table> tables = new HashMap<String, Table>();
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/ConfigFileOptionsBase.java:
```diff
@@ -0,0 +1,53 @@
+        // This option was expected to be present, but it isn't.
+        // Throw error and exit.
+        LOG.error("Invalid Configuration: Expected option " + key + " not found");
+        System.exit(1);
```
It looks like exception is preferable here.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TPCCTableSchemas.java:
```diff
@@ -0,0 +1,159 @@
+            sb.append(", ?");
+        }
+        sb.append(")");
+        return (sb.toString());
```
Unnecessary brackets

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TPCCTableSchemas.java:
```diff
@@ -0,0 +1,159 @@
+                .build()
+            ).collect(Collectors.toMap(TableSchema::name, e -> e));
+    
+    public static String getInsertDml(String tablename) {
```
Current API looks strange.

class `Table` has `getDropDdl` and `getCreateDdl` but doesn't have `getInsertDml`. Is it possible to move `getInsertDml` into `Table` class to avoid blurring of responsibility.

d-uspenskiy commented on src/com/oltpbenchmark/schema/Table.java:
```diff
@@ -0,0 +1,23 @@
+    }
+    
+    // Return the DDL statements to create this table.
+    public abstract List<String> getCreateDdl();
```
Is it necessary to use `List<String>` here? Is not it simper to use single string where multiple statements (if needed) are joined with ';' separator.

deeps1991 replied:
I actually hadn't realized this. Thanks for pointing it out, made this change

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+
+public class TableSchema {
+    private String name;
```
should be `final` and initialized via constructor.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+public class TableSchema {
+    private String name;
+    private final List<Column> columns = new ArrayList<>();
+    private String primaryKey = null;
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+    private String name;
+    private final List<Column> columns = new ArrayList<>();
+    private String primaryKey = null;
+    private String partitionKey = null;
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+
+    static Builder Builder() { return new Builder(); }
+
+    public void setName(String name) {
```
setters are not needed, all fields should be configured in constructor.

deeps1991 replied:
Removed this while addressing your other comment on Builder

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+
+    public String name() { return name; } 
+
+    static class Builder {
```
This version of builder seems very strange because calling of `build` function multiple times always returns same object.

the following common scenario for builders will produce unexpected results

```
Builder b = new Builder();
TableSchema s1 = b.build();
b.column("col1", "int");
TableSchema s2 = b.build();
```

deeps1991 replied:
I looked up another example on Builders and now changed the way the builder works, let me know if this is better. The new Builder always returns a different Schema object.

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    static Builder Builder() { return new Builder(); }
```
what is the purpose of this function?

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultSchemaManager.java:
```diff
@@ -0,0 +1,121 @@
+        execute("CREATE UNIQUE INDEX idx_order ON oorder ((o_w_id,o_d_id) HASH,o_c_id,o_id DESC)");
+
+        if (!db_connection.getAutoCommit()) {
+            db_connection.commit();
```
Is it really a good place to commit the transaction? This function had not start a transaction, why it should commit something?

deeps1991 replied:
I'm not really sure. This code was already there so I had left it as it was and hadn't thought too much into it. I could remove it, but is there something I should test for here?

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultSchemaManager.java:
```diff
@@ -0,0 +1,121 @@
+    
+    @Override
+    public void createSqlProcedures() throws Exception {
+        Statement st = db_connection.createStatement();
```
Statement should be closed after the usage to release resources

```
try (Statement stmt = db_connection.createStatement()) {
```

deeps1991 replied:
Good catch. Fixed

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,128 @@
+ * be replicated and placed.
+ */
+public class GeoPartitionPolicy {
+    boolean isPartitioningEnabled = false;
```
`= false` is redundant

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,52 @@
+        return cloud;
+    }
+
+    public void setCloud(String cloud) {
```
it looks like setters are not used. Please remove them and make all member fields `final`

deeps1991 replied:
For some reason I thought the Gson library needed them, I was mistaken. Removed these

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,52 @@
+    private String cloud;
+    private String region;
+    private String zone;
+    private int min_num_replicas;
```
jave code uses camelCase for multi words naming.

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementPolicy.java:
```diff
@@ -0,0 +1,41 @@
+ */
+ 
+public class PlacementPolicy {
+    int num_replicas;
```
use camelCase for multi words naming

deeps1991 replied:
This class is used by Gson to convert to JSON and use to created tablespace syntax. So I need this field name to have the exact same name used by the SQL syntax. I have added this as a header comment in this file.

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementPolicy.java:
```diff
@@ -0,0 +1,41 @@
+    int num_replicas;
+    List<PlacementBlock> placement_blocks;
+
+    public PlacementPolicy() {
```
redundant constructor, variable can be initialized directly.

```
List<PlacementBlock> placementBlocks = new ArrayList<>();
```

BTW why it has `default` access level. Why not making it `private`?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/SchemaManager.java:
```diff
@@ -0,0 +1,40 @@
+    
+    private static final Logger LOG = Logger.getLogger(SchemaManager.class);
+
+    protected Connection db_connection;
```
`final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/ConfigFileOptionsBase.java:
```diff
@@ -0,0 +1,58 @@
+        return Optional.empty();
+    }
+    
+    protected boolean IsFlagSet(String key) {
```
Not sure we need this helper function as a member. It is possible to create static utility function

```
static boolean isFlagSet(ConfigFileOptionsBase conf, String key) {
   Optional<Boolean> flag = conf.getBoolOpt(key);
   return flag.isPresent() ? flag.get() : false;
}
```

But nevertheless if you think this function should be a member in the `ConfigFileOptionsBase` class please check the name (it should start from lowercase `i`) and update the implementation. It should use existing function `getBoolOpt` or can be this one:

```
static boolean isFlagSet(String key) {
   return xmlConfig.getBoolean(key, false);
}
```

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/WorkloadConfiguration.java:
```diff
@@ -90,6 +95,7 @@ public WorkloadState initializeState(BenchmarkState benchmarkState) {
   private int numberOfPhases = 0;
   private TransactionTypes transTypes = null;
   private int isolationMode = Connection.TRANSACTION_SERIALIZABLE;
+  private GeoPartitionPolicy geoPartitionPolicy;
```
Maybe `final`?

d-uspenskiy commented on src/com/oltpbenchmark/api/Procedure.java:
```diff
@@ -25,7 +25,9 @@
 import java.util.Map.Entry;
 import java.util.Random;
 
+import com.oltpbenchmark.benchmarks.tpcc.TPCCUtil;
```
Look like redundant change?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/api/Procedure.java:
```diff
@@ -25,7 +25,9 @@
 
+import com.oltpbenchmark.benchmarks.tpcc.TPCCUtil;
 import com.oltpbenchmark.jdbc.InstrumentedPreparedStatement;
+import com.oltpbenchmark.util.GeoPartitionPolicy;
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/SchemaManagerFactory.java:
```diff
@@ -0,0 +1,17 @@
+public class SchemaManagerFactory {
+
+    public static SchemaManager getSchemaManager(WorkloadConfiguration workConf, Connection conn) {
+        if (workConf.getGeoPartitioningEnabled()) {
```
What about ternary operator?

```
return workConf.getGeoPartitioningEnabled()
    ? new GeoPartitionedSchemaManager(workConf.getGeoPartitioningPolicy(), conn)
    : new DefaultSchemaManager(conn);
```

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/WorkloadConfiguration.java:
```diff
@@ -358,6 +364,15 @@ public void setMaxLoaderRetries(int maxLoaderRetries) {
+  
+  // Geo partitioned options.
+  public boolean getGeoPartitioningEnabled() {
+      return geoPartitionPolicy.isPartitioningEnabled();
```
Is not it much more simpler to use null for `geoPartitionPolicy` member in case it is not enabled? For now API seems strange. It is possible that `getGeoPartitioningEnabled` return false but `getGeoPartitioningPolicy` returns not null object. How it can be used in this case?
From my point of view single function is enough
```
public GeoPartitionPolicy getGeoPartitioningPolicy() {
    return geoPartitionPolicy;
}
```

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TPCCTableSchemas.java:
```diff
@@ -0,0 +1,151 @@
+                .primaryKey("(w_id)")
+                .partitionKey("(w_id)")
+                .build()
+            ).collect(Collectors.toMap(TableSchema::name, e -> e));
```
It will be good to make resulting map immutable.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/Table.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+public abstract class Table {
+    protected TableSchema schema = null;
```
1. `final`
2. `= null` is redundant

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+}
+
+class TableSchemaBuilder {
+    private String name = null;
```
1. `final`
2. `= null` is redundant
3. Initialize `name` in `TableSchemaBuilder` constructor as it is mandatory field

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+class TableSchemaBuilder {
+    private String name = null;
+    private List<Column> columns = new ArrayList<Column>();
+    private String primaryKey = null;
```
`= null` is redundant

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+    private String name = null;
+    private List<Column> columns = new ArrayList<Column>();
+    private String primaryKey = null;
+    private String partitionKey = null;
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+    private String primaryKey = null;
+    private String partitionKey = null;
+    
+    TableSchemaBuilder name(String name) {
```
Redundant setter, initialize name in constructor instead

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public TableSchema build() {
+        if (name == null || columns.size() == 0) {
```
No need to check `name` here as it is always set via constructor.
Also check columns emptiness with `isEmpty()`

deeps1991 replied:
Changed to isEmpty() but kept the null check for the name because someone may pass null to the constructor.

d-uspenskiy replied:
By following this logic all arguments should be checked for null in java.
In this particular case this name is initialized explicitly, from my point of view null check is not necessary.
In case you are thinking this check should stay you have to make it in constructor.
Also quite common practice in java is to use @NotNull annotation in such cases.

```
class TableSchemaBuilder {
  private final String name;

  public TableSchemaBuilder(@NotNull String name) { 
    this.name = name;
  }
}
```
BTW in case name is checked for null why it is not checked for emptiness?

deeps1991 replied:
I actually did not know about the NotNull annotation. Added it now, thanks Dmitry. And yes I did add an isEmpty check, but that is in the patch that was submitted after this one, that's why it is not visible here.

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultSchemaManager.java:
```diff
@@ -0,0 +1,120 @@
+public class DefaultSchemaManager extends SchemaManager {
+    private static final Logger LOG = Logger.getLogger(DefaultSchemaManager.class);
+    
+    private Map<String, Table> tables = new HashMap<String, Table>();
```
`final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultSchemaManager.java:
```diff
@@ -0,0 +1,120 @@
+
+    @Override
+    public void create() throws SQLException {
+        for (Table t : tables.values()) {
```
Is it possible to use single loop here?


```
for (Table t : tables.values()) {
   execute(t.getDropDdl());
   execute(t.getCreateDdl());
}
```

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultTable.java:
```diff
@@ -0,0 +1,48 @@
+    private final String tablespace;
+    
+    public DefaultTable(TableSchema schema) {
+        super(schema);
```
It is possible to call another constructor via `this(schema, null)`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/defaultschema/DefaultTable.java:
```diff
@@ -0,0 +1,48 @@
+    public String getCreateDdl() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(String.format("CREATE TABLE %s ( ", schema.name()));
+        boolean is_first = true;
```
As the columns is not empty it is possible to simplify this code.
Always add the coma and then trim last symbol.

```
for (Column c : schema.getColumns()) {
   sb.append(String.format("\n%s %s,", c.getName(), c.getDecl()));
}
sb.setLength(sb.length() - 1);
```

d-uspenskiy replied:
It is even better to avoid using format when StringBuilder is used

```
for (Column c : schema.getColumns()) {
   sb.append('\n').append(c.getName()).append(' ').append(c.getDecl()).append(',');
}
sb.setLength(sb.length() - 1);
```

deeps1991 replied:
Nice points, done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/PartitionedTable.java:
```diff
@@ -0,0 +1,100 @@
+import com.oltpbenchmark.util.GeoPartitionPolicy;
+
+public class PartitionedTable extends Table {
+    private GeoPartitionPolicy policy;
```
`final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/PartitionedTable.java:
```diff
@@ -0,0 +1,100 @@
+    private String createPartitionedTable() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(String.format("CREATE TABLE %s ( ", schema.name()));
+        boolean is_first = true;
```
This part of code can be simplified by adding comma always and trimming last symbol at the end.
It is good to have helper function for forming columns description as this code is used in several places.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/PartitionedTable.java:
```diff
@@ -0,0 +1,100 @@
+        StringBuilder sb = new StringBuilder();
+        sb.append(String.format("CREATE TABLE %s%d PARTITION OF %s(", schema.name(), idx, schema.name()));
+
+        boolean is_first = true;
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,128 @@
+ * be replicated and placed.
+ */
+public class GeoPartitionPolicy {
+    boolean isPartitioningEnabled;
```
Why not `private` for all the members?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,128 @@
+
+    public void addTablespacePlacementPolicy(String tablespace, PlacementPolicy policy) {
+        if (tablespaceToPlacementPolicy == null) {
+            tablespaceToPlacementPolicy = new HashMap<String, PlacementPolicy>();
```
Why the `tablespaceToPlacementPolicy` is not initialized with map in constructor? What are the benefits to make lazy initialization?

deeps1991 replied:
Actually earlier didn't want to create these maps upon construction of object because this object could be initialized with isGeoPartitioningEnabled=false. But with your earlier comment I have removed that, so now I've moved this initialization to the constructor.

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,128 @@
+
+    public void addTablespaceForPartition(String tablespace) {
+        if (tablespacesForPartitions == null) {
+            tablespacesForPartitions = new ArrayList<String>();
```
Ditto

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,128 @@
+        tablespacesForPartitions.add(tablespace);
+    }
+
+    public String getTablespaceCreationJson(final String tablespace) {
```
Is the `final` really needed here?

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementPolicy.java:
```diff
@@ -0,0 +1,27 @@
+    private int num_replicas;
+    private List<PlacementBlock> placement_blocks = new ArrayList<PlacementBlock>();
+
+    public void setNumReplicas(int numReplicas) {
```
It looks like it is possible to initialize this field in constructor. Setter can be removed.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementPolicy.java:
```diff
@@ -0,0 +1,27 @@
+ */
+ 
+public class PlacementPolicy {
+    private int num_replicas;
```
`final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementPolicy.java:
```diff
@@ -0,0 +1,27 @@
+ 
+public class PlacementPolicy {
+    private int num_replicas;
+    private List<PlacementBlock> placement_blocks = new ArrayList<PlacementBlock>();
```
`final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,40 @@
+    private final String cloud;
+    private final String region;
+    private final String zone;
+    private final int min_num_replicas;
```
Use camelCase for variables naming.

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,40 @@
+ * This class is used with Gson to generate a JSON to use for tablespace
+ * creation. The field names should not be changed as they should
+ * correspond exactly with the json keys expected by the SQL syntax.
+ */
```
Comment about variable naming is not necessary. It is possible to use cameCase naming for variables and specify alternative naming for gson serialization via `@SerializedName` annotation

deeps1991 replied:
Oh I should have looked for this, totally my bad. Fixed it now, thanks a lot Dmitry!

deeps1991 commented:
Thanks for the detailed review @d-uspenskiy! Addressed your comments in the latest patch.

d-uspenskiy commented on src/com/oltpbenchmark/GeoPartitionedConfigFileOptions.java:
```diff
@@ -0,0 +1,122 @@
+                                    "across partitions.", totalWarehousesAcrossShards,
+                                    numPartitions, numWarehouses));
+        }
+        policy.setNumPartitions(numPartitions);
```
Why this field is not specified in constructor of `GeoPartitionPolicy`? Please check other fields as well and initialize them in constructor in case `GeoPartitionPolicy` class object can't work properly without this fields. Also make all of them `final`

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/geopartitioned/GeoPartitionedSchemaManager.java:
```diff
@@ -0,0 +1,183 @@
+        super(conn);
+        this.geoPartitionPolicy = geoPartitioningPolicy;
+        for (TableSchema t : TPCCTableSchemas.tables.values()) {
+            if (t.name() == "ITEM") {
```
How about this?
```
for (TableSchema t : TPCCTableSchemas.tables.values()) {
  tables.put(t.name(), 
             t.name() == "ITEM" ? new DefaultTable(t, geoPartitioningPolicy.getTablespaceForItemTable()) 
                                : new PartitionedTable(t, geoPartitionPolicy));
}
```

d-uspenskiy replied:
1. `equals` must be used instead of `==`
2. Avoid using of "magic values" use the constant `TPCCConstants.TABLENAME_ITEM` intead

deeps1991 replied:
Done for both comments

d-uspenskiy commented on src/com/oltpbenchmark/util/GeoPartitionPolicy.java:
```diff
@@ -0,0 +1,104 @@
+
+    private String tablespaceForItemTable;
+
+    private List<String> tablespacesForPartitions = new ArrayList<String>();;
```
Nit: redundant `;` at the line end

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/util/PlacementBlock.java:
```diff
@@ -0,0 +1,40 @@
+    @SerializedName(value = "min_num_replicas")
+    private final int minNumReplicas;
+
+    public PlacementBlock(String cloud, String region, String zone, int min_num_replicas) {
```
Please use camelCase for argument

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/ConfigFileOptionsBase.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    protected String getRequiredStringOpt(String key) {
+        if (xmlConfig.containsKey(key) && xmlConfig.getString(key).length() > 0) {
```
How about this?
```
protected String getRequiredStringOpt(String key) {
    return getStringOpt(key).orElseThrow(() -> new InvalidUserConfiguration("Invalid Configuration: Expected option " + key + " not found")).
}
```

deeps1991 replied:
Nice, thanks. Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+    TableSchemaBuilder column(String name, String decl) {
+        this.columns.add(new Column(name, decl));
+        return this;
+    }
```
Nit: Add empty line

deeps1991 replied:
Done

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,65 @@
+        this.primaryKey = primaryKeyDecl;
+        return this;
+    }
+    TableSchemaBuilder partitionKey(String partitionKey) {
```
Ditto

deeps1991 replied:
Done

deeps1991 commented:
Thanks @d-uspenskiy , addressed your latest code review comments

d-uspenskiy commented on src/com/oltpbenchmark/schema/TableSchema.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public TableSchema build() {
+        if (name.isEmpty() || columns.isEmpty()) {
```
All validations of `name` must done in single place. I.e. constructor.

deeps1991 replied:
Done

deeps1991 commented:
Addressed your comment, @d-uspenskiy, thanks for the review!

hbhanawat commented on src/com/oltpbenchmark/GeoPartitionedConfigFileOptions.java:
```diff
@@ -0,0 +1,120 @@
+        
+        // Verify that the total number of warehouses can be evenly divided across partitions.
+        final int numPartitions = getNumberOfPartitions().get();
+        if (numPartitions * numWarehouses != totalWarehousesAcrossShards) {
```
Hey @deeps1991 , we support multiple clients for tpcc and that is why we had totalWarehousesAcrossShards and startWarehouseIdForShard. A client does not know about other clients and it knows only its own startWarehouseIdForShard and numWarehouses.

It does not look like that we are supporting multiple clients per geography. However, the usage of numWarehouses and totalWarehousesAcrossShards in context of geo partitioning has made the things confusing. Two comments here:
1. If multiple clients are not supported for a single geo partitioned region (which I think should be. It is a matter of time before @rkarthik007 would need 100K WH run on geo partitioned cluster), we should throw an exception during config validation.  
2. Use different variable for number of Warehouses in the current partition so that we can add support for 1 in future.

deeps1991 replied:
@hbhanawat The way I thought about it is, every client connects to its own partition. 
So let's say you have 100K warehouses and you want to have (just as an example) 20 clients in total. Then you can create 20 partitions, and each partition will house 5K warehouses. You can spread these clients across geographies however you want, so I'm not completely sure what you mean when you say that multiple clients are not supported for single geo partitioned run.
Am I misunderstanding something here? AFAIK totalWarehousesAcrossShards would be 100K, numWarehouses=5k and numPartitions=20 (and total number of clients =20) in the above example. If you have geo distributed cluster of 2 regions you can 10 partitions/clients in each region, or is something wrong above.
Do you mean to say that we should not tie number of clients with number of partitions?

deeps1991 commented:
@hbhanawat I made a change to remove the dependency between number of clients and number of partitions. I made the validation check less restrictive - it no longer checks that the client is exactly matching partition boundaries, rather it only checks whether the client is operating on some range of warehouses that are completely within one partition (this is for the FK checks to work). This way you could split into say 3 partitions, but have like 30 clients connect, just as long as each client operates on a range of warehouses that lies completely within one partition.
Let me know if this addresses all your concerns.

hbhanawat commented:
Have reviewed the last commit that removes the dependency between number of clients and number of partitions. That looks good.
